Added SocketCAN driver implementation to the tiva chip, modified the …

dvosahlik opened the pull request:
## Summary
Added implementation of SocketCAN driver to the Tiva chips.

## Impact
Modified the configuration for the Tiva chips, added source files for the SocketCAN driver implementation. Modified the boards/arm/tiva/tm4c1294-launchpad/src/tm4c_can.c to use the new SocketCAN driver.
## Testing
The implementation was tested on the EK-TM4C1294XL launchpad connected to the PEAK 2 USB converter. It was tested using the cansend and candump applications. It was tested with the bugfix proposed in #8293

xiaoxiang781216 commented on boards/arm/tiva/tm4c1294-launchpad/src/tm4c_can.c:
```diff
@@ -115,11 +131,25 @@ int tm4c_can_setup(void)
   return OK;
 
-configgpio_error:
+  configgpio_error:
```
revert the change?

dvosahlik replied:
done

acassis commented:
@dvosahlik you cannot add NET_CAN_HAVE_ERRORS for each chip, it depends on Socket CAN enabled for that chip, please take a look at STM32_CAN_SOCKET and STM32_FDCAN_SOCKET to see how it is done on STM32

dvosahlik commented:
> @dvosahlik you cannot add NET_CAN_HAVE_ERRORS for each chip, it depends on Socket CAN enabled for that chip, please take a look at STM32_CAN_SOCKET and STM32_FDCAN_SOCKET to see how it is done on STM32

@acassis I added that for all the tiva chips, that have CAN peripheral and for which the SocketCAN is implemented in this PR.

acassis commented:
> > @dvosahlik you cannot add NET_CAN_HAVE_ERRORS for each chip, it depends on Socket CAN enabled for that chip, please take a look at STM32_CAN_SOCKET and STM32_FDCAN_SOCKET to see how it is done on STM32
> 
> @acassis I added that for all the tiva chips, that have CAN peripheral and for which the SocketCAN is implemented in this PR.

Yes, but NET_CAN_HAVE_ERRORS only exist when users enable the CAN. The chip itself having CAN doesn't imply that users will enable it. So, it should be enabled by TIVA_CAN, not by the chip.

dvosahlik commented:
> > > @dvosahlik you cannot add NET_CAN_HAVE_ERRORS for each chip, it depends on Socket CAN enabled for that chip, please take a look at STM32_CAN_SOCKET and STM32_FDCAN_SOCKET to see how it is done on STM32
> > 
> > 
> > @acassis I added that for all the tiva chips, that have CAN peripheral and for which the SocketCAN is implemented in this PR.
> 
> Yes, but NET_CAN_HAVE_ERRORS only exist when users enable the CAN. The chip itself having CAN doesn't imply that users will enable it. So, it should be enabled by TIVA_CAN, not by the chip.

Ok, I misinterpreted the meaning of NET_CAN_HAVE_ERRORS then. I will modify the configuration file then.

dvosahlik commented:
@acassis I followed the STM32 example and moved the NET_CAN_HAVE_ERRORS to be selected only when the socketCAN driver is chosen.
@xiaoxiang781216 I reverted the change you wanted me to revert.
I also updated the usage of macros in the `tivacan_handle_errors` function. I use nuttx/can.h defined macros instead of the proprietary defined macros, which are used in the character device driver.

hartmannathan commented on arch/arm/src/tiva/common/tiva_sock_can.c:
```diff
@@ -0,0 +1,2371 @@
+  int      errcnt;
+  int      prev_errcnt;
+
+  memset(frame, 0, sizeof(struct can_frame));
```
```suggestion
  memset(frame, 0, sizeof(*frame));
```

Just to future-proof the code against structural changes...

hartmannathan commented on arch/arm/src/tiva/common/tiva_sock_can.c:
```diff
@@ -0,0 +1,2371 @@
+  if (tivacan_get_txrq32(canmod->base) & ~tivacan_get_nwda32(canmod->base)
+  && cansts & TIVA_CAN_STS_RXOK)
+    {
+      /* Received a message successfually but sending a message failed */
```
```suggestion
      /* Received a message successfully but sending a message failed */
```

hartmannathan commented on arch/arm/src/tiva/common/tiva_sock_can.c:
```diff
@@ -0,0 +1,2371 @@
+  canmod->dev.d_private = canmod;
+
+  /* Put the interface in the down state.  This usually amounts to resetting
+   * the device and/or calling stm32can_ifdown().
```
```suggestion
   * the device and/or calling tivacan_ifdown().
```
